File: next_permutation.py
# https://leetcode.com/problems/next-permutation/description/

class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        idx = -1
        n = len(nums)
        for i in range(n - 2, -1, -1):
            if nums[i] < nums[i + 1]:
                idx = i
                break
        if idx == -1: self.rev(nums, 0, n - 1)
        else:
            for i in range(n - 1, idx - 1, -1):
                if nums[i] > nums[idx]:
                    t = nums[i]
                    nums[i] = nums[idx]
                    nums[idx] = t
                    break
            self.rev(nums, idx + 1, n - 1)
    def rev(self, nums: List[int], a: int, b: int) -> None:
        l, r = a, b
        while l < r:
            t = nums[l]
            nums[l] = nums[r]
            nums[r] = t
            l += 1
            r -= 1
        return

File: next_Permutation.java
// https://leetcode.com/problems/next-permutation/description/

class Solution {
    public void nextPermutation(int[] nums) {
        int ind = -1;
        int n = nums.length;
        for(int i = n - 2; i >= 0; i--) {
            if(nums[i] < nums[i + 1]) {
                ind = i;
                break;
            }
        }
        if(ind == -1) {
            rev(nums, 0, n - 1);
        } else {
            for(int i = n - 1; i >= ind; i--) {
                if(nums[i] > nums[ind]) {
                    int temp = nums[i];
                    nums[i] = nums[ind];
                    nums[ind] = temp;
                    break;
                }
            }
            rev(nums, ind + 1, n - 1);
        }
    }
    public static void rev(int[] nums, int a, int b) {
        int l = a, r = b;
        while(l < r) {
            int t = nums[l];
            nums[l] = nums[r];
            nums[r] = t;
            l++;
            r--;
        }
        return;
    }
}
